### ISSUE TITLE
mismatched data in process list, in Windows 7

### ISSUE THREAD

**Issue by jn1** [jezhumble/javasysmon#9]
As reported in the google group.  This is with Windows 7, Oracle JDK 1.7.0_25, javasysmon-0.3.3.jar.

In the output below, notice that the process name and process command string don't always match up; for some records it does, and for some records it doesn't.  As just one example, notice that the record containing "calc.exe" as a command string, says that the process name is "bash.exe".

Code:

```
Monitor monitor = new JavaSysMon();
for (ProcessInfo process : monitor.processTable()) {
  int pid = process.getPid();
  String name = process.getName();
  String command = process.getCommand();
  System.out.println(String.format("%7s %20s \"%s\"", pid, name, command));
}
```

Output:

```
  0            <unknown> "<unknown>"
  4            <unknown> "<unknown>"
328            <unknown> "<unknown>"
420            <unknown> "<unknown>"
488            <unknown> "<unknown>"
496            <unknown> "<unknown>"
548            <unknown> "<unknown>"
556            <unknown> "<unknown>"
568            <unknown> "<unknown>"
628            <unknown> "<unknown>"
712            <unknown> "<unknown>"
784            <unknown> "<unknown>"
824            <unknown> "<unknown>"
932            <unknown> "<unknown>"
964            <unknown> "<unknown>"
```

   1004            <unknown> "<unknown>"
    808            <unknown> "<unknown>"
   1112            <unknown> "<unknown>"
   1136            <unknown> "<unknown>"
   1240            <unknown> "<unknown>"
   1312            <unknown> "<unknown>"
   1544            <unknown> "<unknown>"
   1572            <unknown> "<unknown>"
   1740            <unknown> "<unknown>"
   1764            <unknown> "<unknown>"
   1828            <unknown> "<unknown>"
   1948            <unknown> "<unknown>"
   1204            <unknown> "<unknown>"
   2024            <unknown> "<unknown>"
   1144            <unknown> "<unknown>"
   1092            <unknown> "<unknown>"
   2764            <unknown> "<unknown>"
   2848            <unknown> "<unknown>"
    996            <unknown> "\Device\HarddiskVolume2\Windows\System32\taskhost.exe"
   2172            <unknown> "\Device\HarddiskVolume2\Windows\System32\dwm.exe"
    484            <unknown> "\Device\HarddiskVolume2\Windows\explorer.exe"
   2728            <unknown> "\Device\HarddiskVolume2\Windows\explorer.exe"
    728          msnmsgr.exe "\Device\HarddiskVolume2\Program Files (x86)\Windows Live\Messenger\msnmsgr.exe"
   2448         trillian.exe "\Device\HarddiskVolume2\Program Files (x86)\Trillian\trillian.exe"
   2200         smax4pnp.exe "\Device\HarddiskVolume2\Program Files (x86)\Analog Devices\Core\smax4pnp.exe"
   2428         smax4pnp.exe "\Device\HarddiskVolume2\Program Files (x86)\Analog Devices\Core\smax4pnp.exe"
   2032            ccApp.exe "\Device\HarddiskVolume2\Program Files (x86)\Common Files\Symantec Shared\ccApp.exe"
   3328           hqtray.exe "\Device\HarddiskVolume2\Program Files (x86)\VMware\VMware Player\hqtray.exe"
   4048           hqtray.exe "\Device\HarddiskVolume2\Program Files (x86)\VMware\VMware Player\hqtray.exe"
   3928           wlcomm.exe "\Device\HarddiskVolume2\Program Files (x86)\Windows Live\Contacts\wlcomm.exe"
   4952           wlcomm.exe "\Device\HarddiskVolume2\Program Files (x86)\Windows Live\Contacts\wlcomm.exe"
   6020           wlcomm.exe "\Device\HarddiskVolume2\Program Files (x86)\Windows Live\Contacts\wlcomm.exe"
   7152           wlcomm.exe "\Device\HarddiskVolume2\Program Files\Eclipse\eclipse.exe"
   6008           wlcomm.exe "\Device\HarddiskVolume2\Windows\System32\javaw.exe"
   6844      thunderbird.exe "\Device\HarddiskVolume2\Program Files (x86)\Mozilla Thunderbird\thunderbird.exe"
   4148          firefox.exe "\Device\HarddiskVolume2\Program Files (x86)\Mozilla Firefox\firefox.exe"
   2368 plugin-container.exe "\Device\HarddiskVolume2\Program Files (x86)\Mozilla Firefox\plugin-container.exe"
   3184 plugin-container.exe "\Device\HarddiskVolume2\Windows\System32\cmd.exe"
    640 plugin-container.exe "\Device\HarddiskVolume2\Windows\System32\conhost.exe"
   7024          TextPad.exe "\Device\HarddiskVolume2\Program Files (x86)\TextPad 5\TextPad.exe"
   4772             wish.exe "\Device\HarddiskVolume2\Program Files (x86)\Git\bin\wish.exe"
   4512             wish.exe "\Device\HarddiskVolume2\Program Files\Eclipse\eclipse.exe"
   6964             wish.exe "\Device\HarddiskVolume2\Windows\System32\javaw.exe"
   4164 plugin-container.exe "\Device\HarddiskVolume2\Program Files (x86)\Mozilla Firefox\plugin-container.exe"
   5912 plugin-container.exe "\Device\HarddiskVolume2\Windows\splwow64.exe"
   4396 plugin-container.exe "\Device\HarddiskVolume2\Windows\System32\cmd.exe"
   2484 plugin-container.exe "\Device\HarddiskVolume2\Windows\System32\conhost.exe"
   3288             bash.exe "\Device\HarddiskVolume2\programs\cygwin\bin\bash.exe"
   6956             bash.exe "\Device\HarddiskVolume2\Windows\System32\cmd.exe"
   4812             bash.exe "\Device\HarddiskVolume2\Windows\System32\conhost.exe"
   6104             bash.exe "\Device\HarddiskVolume2\Program Files\Java\jdk1.6.0_21\bin\java.exe"
   1152            javaw.exe "\Device\HarddiskVolume2\dev\sdf\tools\jdk1.6.0_25\win32\bin\javaw.exe"
    644            javaw.exe "\Device\HarddiskVolume2\dev\sdf\tools\jdk1.6.0_25\win32\bin\javaw.exe"
   6992            javaw.exe "\Device\HarddiskVolume2\dev\sdf\tools\jdk1.6.0_25\win32\bin\javaw.exe"
   5192            javaw.exe "\Device\HarddiskVolume2\Windows\System32\cmd.exe"
   1776            javaw.exe "\Device\HarddiskVolume2\Windows\System32\conhost.exe"
   5928             bash.exe "\Device\HarddiskVolume2\programs\cygwin\bin\bash.exe"
   8468             bash.exe "\Device\HarddiskVolume2\programs\cygwin\bin\bash.exe"
   7392             bash.exe "\Device\HarddiskVolume2\Windows\System32\calc.exe"

**Comment by jn1** [comment]
Github garbled part of the output.  I just figured how to quote it, so here it is again.  It's slightly different, as it's from a different run.  Note that I only have one instance of calc.exe running, but I see it twice here.

```
      0            <unknown> "<unknown>"
      4            <unknown> "<unknown>"
    328            <unknown> "<unknown>"
    420            <unknown> "<unknown>"
    488            <unknown> "<unknown>"
    496            <unknown> "<unknown>"
    548            <unknown> "<unknown>"
    556            <unknown> "<unknown>"
    568            <unknown> "<unknown>"
    628            <unknown> "<unknown>"
    712            <unknown> "<unknown>"
    784            <unknown> "<unknown>"
    824            <unknown> "<unknown>"
    932            <unknown> "<unknown>"
    964            <unknown> "<unknown>"
   1004            <unknown> "<unknown>"
    808            <unknown> "<unknown>"
   1112            <unknown> "<unknown>"
   1136            <unknown> "<unknown>"
   1240            <unknown> "<unknown>"
   1312            <unknown> "<unknown>"
   1544            <unknown> "<unknown>"
   1572            <unknown> "<unknown>"
   1740            <unknown> "<unknown>"
   1764            <unknown> "<unknown>"
   1828            <unknown> "<unknown>"
   1948            <unknown> "<unknown>"
   1204            <unknown> "<unknown>"
   2024            <unknown> "<unknown>"
   1144            <unknown> "<unknown>"
   1092            <unknown> "<unknown>"
   2764            <unknown> "<unknown>"
   2848            <unknown> "<unknown>"
    996            <unknown> "\Device\HarddiskVolume2\Windows\System32\taskhost.exe"
   2172            <unknown> "\Device\HarddiskVolume2\Windows\System32\dwm.exe"
    484            <unknown> "\Device\HarddiskVolume2\Windows\explorer.exe"
   2728            <unknown> "\Device\HarddiskVolume2\Windows\explorer.exe"
    728          msnmsgr.exe "\Device\HarddiskVolume2\Program Files (x86)\Windows Live\Messenger\msnmsgr.exe"
   2448         trillian.exe "\Device\HarddiskVolume2\Program Files (x86)\Trillian\trillian.exe"
   2200         smax4pnp.exe "\Device\HarddiskVolume2\Program Files (x86)\Analog Devices\Core\smax4pnp.exe"
   2428         smax4pnp.exe "\Device\HarddiskVolume2\Program Files (x86)\Analog Devices\Core\smax4pnp.exe"
   2032            ccApp.exe "\Device\HarddiskVolume2\Program Files (x86)\Common Files\Symantec Shared\ccApp.exe"
   3328           hqtray.exe "\Device\HarddiskVolume2\Program Files (x86)\VMware\VMware Player\hqtray.exe"
   4048           hqtray.exe "\Device\HarddiskVolume2\Program Files (x86)\VMware\VMware Player\hqtray.exe"
   3928           wlcomm.exe "\Device\HarddiskVolume2\Program Files (x86)\Windows Live\Contacts\wlcomm.exe"
   4952           wlcomm.exe "\Device\HarddiskVolume2\Program Files (x86)\Windows Live\Contacts\wlcomm.exe"
   6020           wlcomm.exe "\Device\HarddiskVolume2\Program Files (x86)\Windows Live\Contacts\wlcomm.exe"
   7152           wlcomm.exe "\Device\HarddiskVolume2\Program Files\Eclipse\eclipse.exe"
   6008           wlcomm.exe "\Device\HarddiskVolume2\Windows\System32\javaw.exe"
   6844      thunderbird.exe "\Device\HarddiskVolume2\Program Files (x86)\Mozilla Thunderbird\thunderbird.exe"
   4148          firefox.exe "\Device\HarddiskVolume2\Program Files (x86)\Mozilla Firefox\firefox.exe"
   2368 plugin-container.exe "\Device\HarddiskVolume2\Program Files (x86)\Mozilla Firefox\plugin-container.exe"
   3184 plugin-container.exe "\Device\HarddiskVolume2\Windows\System32\cmd.exe"
    640 plugin-container.exe "\Device\HarddiskVolume2\Windows\System32\conhost.exe"
   7024          TextPad.exe "\Device\HarddiskVolume2\Program Files (x86)\TextPad 5\TextPad.exe"
   4772             wish.exe "\Device\HarddiskVolume2\Program Files (x86)\Git\bin\wish.exe"
   4512             wish.exe "\Device\HarddiskVolume2\Program Files\Eclipse\eclipse.exe"
   6964             wish.exe "\Device\HarddiskVolume2\Windows\System32\javaw.exe"
   4164 plugin-container.exe "\Device\HarddiskVolume2\Program Files (x86)\Mozilla Firefox\plugin-container.exe"
   5912 plugin-container.exe "\Device\HarddiskVolume2\Windows\splwow64.exe"
   4396 plugin-container.exe "\Device\HarddiskVolume2\Windows\System32\cmd.exe"
   2484 plugin-container.exe "\Device\HarddiskVolume2\Windows\System32\conhost.exe"
   3288             bash.exe "\Device\HarddiskVolume2\programs\cygwin\bin\bash.exe"
   6956             bash.exe "\Device\HarddiskVolume2\Windows\System32\cmd.exe"
   4812             bash.exe "\Device\HarddiskVolume2\Windows\System32\conhost.exe"
   6104             bash.exe "\Device\HarddiskVolume2\Program Files\Java\jdk1.6.0_21\bin\java.exe"
   1152            javaw.exe "\Device\HarddiskVolume2\dev\sdf\tools\jdk1.6.0_25\win32\bin\javaw.exe"
    644            javaw.exe "\Device\HarddiskVolume2\dev\sdf\tools\jdk1.6.0_25\win32\bin\javaw.exe"
   6992            javaw.exe "\Device\HarddiskVolume2\dev\sdf\tools\jdk1.6.0_25\win32\bin\javaw.exe"
   5192            javaw.exe "\Device\HarddiskVolume2\Windows\System32\cmd.exe"
   1776            javaw.exe "\Device\HarddiskVolume2\Windows\System32\conhost.exe"
   5928             bash.exe "\Device\HarddiskVolume2\programs\cygwin\bin\bash.exe"
   7392             bash.exe "\Device\HarddiskVolume2\Windows\System32\calc.exe"
   8708             bash.exe "\Device\HarddiskVolume2\Windows\System32\calc.exe"
```

**Comment by watsonmw** [comment]
It's because the process name isn't cleared when you don't have permission to get it.

I'm guessing the return value of GetModuleBaseName() should be checked and 'process_name' nulled out if the module name can't be determined.

https://github.com/jezhumble/javasysmon/blob/master/src/main/c/windows/javasysmon.c

**Comment by watsonmw** [comment]
Added a potential fix in #22
